src/common/download.cpp

```diff
@@ -40,27 +41,86 @@
 
 namespace tools
 {
-  static bool download_thread(const std::string &path, const std::string &url)
+  struct download_thread_control
   {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool stop;
+    bool stopped;
+    bool success;
+    std::shared_ptr<boost::thread> thread;
+    boost::mutex mutex;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), stop(false), stopped(false), success(false) {}
+    void lock() { mutex.lock(); }
+    void unlock() { mutex.unlock(); }
+  };
+
+  static void download_thread(download_thread_control *control)
+  {
+    static std::atomic<unsigned int> thread_id(0);
+
+    MLOG_SET_THREAD_NAME("DL" + std::to_string(thread_id++));
+
+    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&control](){ delete control; });
```

Comment:
@fluffypony unfortunately I have added a few things into the `epee` namespace. Easy to spot and liberate eventually due to copyright differences in the file. Or maybe the liberation can be done now - is calling from `epee` into the monero src verboten (a few sneaky calls to crypto::rand exist) ?

@moneromooo-monero  A lock "inside" of the pointer cannot prevent its own destruction because the lock cannot be held during the free call. If the caller tries to "free" the pointer, it can be made non-blocking by allowing the thread to clean up the memory (like this implementation). However, the caller usually wants blocking behavior if the callback has a reference or raw pointer. For example, `core` _requires_ blocking behavior, otherwise the download thread could invoke a callback to a freed `core` pointer. OR that callback needs to cleanup the memory for `core` (which is probably more complicated than you want).